File: Java/compoundlab/src/main/java/com/bbd/abstractFactoryPattern/RubberDuck.java
package com.bbd.abstractFactoryPattern;

public class RubberDuck implements IQuackable
{
	private String _name = "Rubber Duck";
	IQuackObservable _observable;

	RubberDuck()
	{
		this._observable = new Observable(this);
	}

	RubberDuck(IQuackObservable observable)
	{
		this._observable = observable;
	}

	public String quack()
	{
		notifyObservers();
		return ("Rubber duck quacks");
	}

	public void registerObserver(IObserver observer)
	{
		this._observable.registerObserver(observer);
	}

	public String notifyObservers()
	{
		return (this._observable.notifyObservers());
	}

	@Override
	public String getName() { return  this._name;}
}

File: Scala/compoundlab/src/main/scala/RubberDuck.scala
class RubberDuck(observable: Option[IQuackableObservable] = None) extends IQuackable {

	private val name = "Rubber Duck"
	private val _observable : IQuackableObservable = observable getOrElse new Observable(this)


	override def registerObserver(observer : IObserver) : Unit = {
		this._observable.registerObserver(observer)
	}

	override def notifyObservers: String = {
		this._observable.notifyObservers
	}

	override def quack : String = {
		notifyObservers
		"Rubber duck quacked"
	}

	override def getName: String = {
		this.name
	}
}
